### ISSUE TITLE
Transformer decoder is giving the same out for every example

### ISSUE THREAD

**Issue by santhoshkolloju** [asyml/texar#79]
I have been using texar library to solve summarization problem. I have replaced encoder part with Bert and decode is Transformer Decoder with beam size 5.
My loss got decreased from 11 to around 3 after 50 k iteration but when I try to infer on test data it gives me same output for every example any thoughts on this?
Link to the code:
https://github.com/santhoshkolloju/bert_summ

**Comment by ZhitingHu** [comment]
By my experience, the transformer decoder works well on other tasks. I personally haven't tried BERT as an encoder for generation.

What summarization data are you using?

Have you tried replacing transformer decoder with an RNN decoder for a quick comparison?

**Comment by santhoshkolloju** [comment]
I am using CNN data for building the model. I have not tried RNN decoder will try to do that .
But can you please check the code once in the given link if something is wrong.

**Comment by santhoshkolloju** [comment]
I was able to debug the the problem. Bert weights got disturbed and it was giving similar embeddings irrespective of example I am passing so the context of the sentence at the decoder is not correctly captured. Is there a way where I can set trainable as false for Bert based encoder directly?

**Comment by ZhitingHu** [comment]
You can use `tf.stop_gradient` to disable gradient backpropagation to BERT, or specify the `variables` argument when calling [`tf.contrib.layers.optimize_loss`](https://github.com/asyml/texar/blob/master/examples/bert/bert_classifier_main.py#L193) to exclude BERT variables

**Comment by amudha1** [comment]
I am still facing the same issue.
Can you please elaborate on how you solved it?

**Comment by amudha1** [comment]
Is this the correct way?
hparams=tf.stop_gradient(hparams)

**Comment by amudha1** [comment]
This seems to be working:
encoder_output = tf.stop_gradient(encoder_output)
Is this the right way ?

**Comment by santhoshkolloju** [comment]
allvars = tf get_trainable_variables()
nonBert =[v for v in allvars if 'bert' not in v]

train_op = tx.core.get_train_op(
mle_loss,
learning_rate=learning_rate,
variables=non Bert,
global_step=global_step,
hparams=opt)
